src/UnionFind.hs

```diff
@@ -0,0 +1,201 @@
+  if ancestor1 == ancestor2 then
+    unionFind2
+  else
+    let onUnionFind2sMap update = UnionFind $ update $ getMap $ unionFind2 in
+    case (join maybeValue1, join maybeValue2) of
+    (Nothing, Nothing) ->
+      onUnionFind2sMap $ Map.insert key1 (ChildOf key2)
+                       . Map.insert key2  Ancestor
+    (Just _, Nothing) ->
+      onUnionFind2sMap $ Map.insert key2 (ChildOf key1)
+    (Nothing, Just _) ->
+      onUnionFind2sMap $ Map.insert key1 (ChildOf key2)
+    (Just value1, Just value2) ->
+      -- FIXME: Implement the optimisation that choses which key should be the
+      -- other's child by keeping track of the size of the equivalence classes.
+      onUnionFind2sMap $ Map.insert key1 (ChildOf key2)
+                       . Map.insert key2 (AncestorWith $ merge value1 value2)
+
+-- | Same as @union@ for trivial cases where one knows for sure that the keys
+-- are not in the same equivalence class (or one is absent).
+--
+trivialUnion :: Ord key =>
+  key ->
+  key ->
+  UnionFind key value ->
+  UnionFind key value
+trivialUnion = union (\_ _ -> error "union was not trivial")
+
+-- | @insert merge unionFind key value@ adds a binding from @key@ to @value@ in
+-- the structure. If @key@ is already known in the structure, then @merge@ is
+-- called to reconcile the known value and the new one. @merge@ receives the new
+-- value as first argument.
+--
+insert :: Ord key =>
+  (value -> value -> value) ->
+  key ->
+  value ->
+  UnionFind key value ->
+  UnionFind key value
+insert merge key value unionFind =
+  let (unionFind', ancestor, maybeValue) = find key unionFind in
+  let newValue = maybe value (merge value) (join maybeValue) in
+  UnionFind $ Map.insert ancestor (AncestorWith newValue) $ getMap unionFind'
+
+-- | Same as @insert@ for trivial cases where one knows for sure that the key is
+-- not already in the structure.
+--
+trivialInsert :: Ord key =>
+  key ->
+  value ->
+  UnionFind key value ->
+  UnionFind key value
+trivialInsert = insert (\_ _ -> error "insert was not trivial")
+
+-- | It is guaranteed that the right-hand side of a pair in the list of
+-- equalities is the representant of an equivalence class, and therefore it
+-- occurs on the left-hand side in the list of bindings.
+--
+toLists :: Ord key => UnionFind key value -> (UnionFind key value, [(key, key)], [(key, value)])
+toLists unionFind =
+  foldl
+    (\(unionFind, equalities, bindings) (key, binding) ->
+        case binding of
+          ChildOf _ ->
+            -- FIXME: this is dropping the optimised union-find.
+            let (unionFind', ancestor, _) = find key unionFind in
+              (unionFind', (key, ancestor) : equalities, bindings)
+          Ancestor ->
+            (unionFind, equalities, bindings)
+          AncestorWith value ->
+            (unionFind, equalities, (key, value) : bindings))
+    (unionFind, [], [])
+    (Map.toList $ getMap unionFind)
+
+join :: Maybe (Maybe value) -> Maybe value
```

Comment:
This is just a helper (that would actually work for any monad) but I don't think it's present in any library that we are using? And it feels a bit dumb to add a dependency just for this.

Reply:
I think this is just `join` in `Control.Monad`. Unless I'm misunderstanding something, we should probably just use that.

Reply:
This is definitely `join` in `Control.Monad`. For some reason, it didn't show up in Hoogle; I must've done something fishy.